### PR TITLE
update: add JVM records

### PR DESCRIPTION
**Pull request by koshachy**

### CODE REVIEW

File: docs/topics/jvm/jvm-records.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+[//]: # (title: Records)
+
+> JVM records feature is [Experimental](components-stability.md). It may be dropped or changed at any time. Use it only for evaluation purposes. We would appreciate your feedback on it in [YouTrack](https://youtrack.jetbrains.com/issue/KT-42430).
+>
+{type="warning"}
+
+_Records_ are Java specific classes for storing immutable data. Records carry a fixed set of values – the _records components_.
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:7]: Do you mean `Java-specific classes` or `specific Java classes`? It not clear for me now, maybe it's better to rephrase.
I'd also add a link to the Java doc page about records.

File: docs/topics/jvm/jvm-records.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+As compared to regular Java classes, they have a concise syntax in Java and save you from writing boilerplate code:
+
+```java
+public record Person (String name, int age) {}
+```
+
+The compiler automatically generates a final class inherited from [java.lang.Record](https://download.java.net/java/early_access/jdk16/docs/api/java.base/java/lang/Record.html) with the following members:
+* a private final field for each record component
+* a public constructor
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:16]: With parameters for all fields, I suppose? Let's be precise here.

File: docs/topics/jvm/jvm-records.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+* a set of methods to implement structural equality: `equals()`, `hashCode()`, `toString()`
+* a public method for reading the component
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:18]: I don't get what is `the component` here.

File: docs/topics/jvm/jvm-records.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+> JVM records feature is [Experimental](components-stability.md). It may be dropped or changed at any time. Use it only for evaluation purposes. We would appreciate your feedback on it in [YouTrack](https://youtrack.jetbrains.com/issue/KT-42430).
+>
+{type="warning"}
+
+_Records_ are Java specific classes for storing immutable data. Records carry a fixed set of values – the _records components_.
+As compared to regular Java classes, they have a concise syntax in Java and save you from writing boilerplate code:
+
+```java
+public record Person (String name, int age) {}
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:11]: If a page contains samples in both Kotlin and Java, please start Java snippets with the `// Java` comment as we do in [Calling Kotlin from Java](https://kotlinlang.org/docs/java-to-kotlin-interop.html), for example.

File: docs/topics/jvm/jvm-records.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+This JVM-specific annotation enables generating:
+
+* the record components corresponding to the class properties in the class file
+* the property accessor methods named according to the Java record naming convention
+
+The data class provides `equals()`, `hashCode()`, `toString()` method implementations.
+
+### Requirements
+
+To declare a data class with the `@JvmRecord` annotation, it should meet the following requirements:
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:55]: Seems grammatically incorrect: `To declare..., it should meet`

File: docs/topics/jvm/jvm-records.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+* The class shall be in a module that targets JVM 16 bytecode (or 15 if  the`-Xjvm-enable-preview` compiler option is enabled).
+* The class cannot explicitly inherit any other class (including `Any`) because all JVM records implicitly inherit `java.lang.Record`.
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:58]: Is this about direct `extends` relation or implementing interfaces is prohibited as well?
The current wording seems a bit vague, I'd make it more concrete.

File: docs/topics/jvm/jvm-records.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+* The class cannot declare any additional state – properties with backing fields – except those initialized from the corresponding primary constructor parameters.
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:59]: Do we really need `additional state` here? Seems like this complicates the explanation without an obvious value.

File: docs/topics/jvm/jvm-records.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+* The class cannot declare any mutable state – mutable properties with backing fields.
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:60]: Same as above - is `mutable state` necessary?

File: docs/topics/jvm/jvm-records.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+* The class cannot be local.
+* The primary constructor of the class must be as visible as the class itself.
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:62]: Nitpick: I like `must have the same visibility` more. It's up to you anyway :)

File: docs/topics/jvm/jvm-records.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
+[//]: # (title: Records)
+
+> JVM records feature is [Experimental](components-stability.md). It may be dropped or changed at any time. Use it only for evaluation purposes. We would appreciate your feedback on it in [YouTrack](https://youtrack.jetbrains.com/issue/KT-42430).
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:3]: In our guidelines we have a special warning for experimental features with opt-in, see [here](https://docs.google.com/document/d/1zmjZvQHsvkohMZCQTpv64gPDZEpUOQwYqQ-YH601oWc/edit#heading=h.ty9tig7yrvtm) or an example on the [Sealed interfaces](https://kotlinlang.org/docs/sealed-classes.html#sealed-interfaces) doc.

**Review by p7nov — COMMENTED** [review]
Please see my comments.

File: docs/topics/jvm/jvm-records.md (hunk 10 above)
Comment (koshachy) [reply docs/topics/jvm/jvm-records.md]: How fast do they change! %)
I took the one from Sealed interfaces

File: docs/topics/jvm/jvm-records.md (hunk 9 above)
Comment (koshachy) [reply docs/topics/jvm/jvm-records.md:64]: Copyeditors approved this variant.

File: docs/topics/jvm/jvm-records.md (hunk 7 above)
Comment (koshachy) [reply docs/topics/jvm/jvm-records.md]: Seems redundant.

File: docs/topics/jvm/jvm-records.md (hunk 8 above)
Comment (koshachy) [reply docs/topics/jvm/jvm-records.md]: Same.

File: docs/topics/jvm/jvm-records.md (hunk 6 above)
Comment (koshachy) [reply docs/topics/jvm/jvm-records.md]: I thought the difference in `inherit` vs `implement`. 
I've added the remark.

File: docs/topics/jvm/jvm-records.md (hunk 5 above)
Comment (koshachy) [reply docs/topics/jvm/jvm-records.md]: I took the version from copyeditors (but it still has the same structure).
What do you think about that?

File: docs/topics/jvm/jvm-records.md (hunk 4 above)
Comment (koshachy) [reply docs/topics/jvm/jvm-records.md:13]: Fixed, thx!

File: docs/topics/jvm/jvm-records.md (hunk 3 above)
Comment (koshachy) [reply docs/topics/jvm/jvm-records.md]: Add `record component`

File: docs/topics/jvm/jvm-records.md (hunk 2 above)
Comment (koshachy) [reply docs/topics/jvm/jvm-records.md]: Sure, I took your suggestion.

File: docs/topics/jvm/jvm-records.md (hunk 1 above)
Comment (koshachy) [reply docs/topics/jvm/jvm-records.md]: Rephrased to be more precise.

File: docs/topics/jvm/jvm-records.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+_Records_ are [classes](https://openjdk.java.net/jeps/395) in Java for storing immutable data. Records carry a fixed set of values – the _records components_.
+They have a concise syntax in Java and save you from having to write boilerplate code:
+
+```java
+// Java
+public record Person (String name, int age) {}
+```
+
+The compiler automatically generates a final class inherited from [java.lang.Record](https://download.java.net/java/early_access/jdk16/docs/api/java.base/java/lang/Record.html) with the following members:
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:16]:
```suggestion
The compiler automatically generates a final class inherited from [`java.lang.Record`](https://download.java.net/java/early_access/jdk16/docs/api/java.base/java/lang/Record.html) with the following members:
```

File: docs/topics/jvm/jvm-records.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+* a public constructor with parameters for all fields
+* a set of methods to implement structural equality: `equals()`, `hashCode()`, `toString()`
+* a public method for reading each record component
+
+Records are very similar to Kotlin [data classes](data-classes.md).
+
+## Using Java records from Kotlin code
+
+You can use record classes with components that are declared in Java the same way you would use classes with properties in Kotlin.
+To access the record component, just use its name like you do for [Kotlin properties](https://kotlinlang.org/docs/properties.html):
```
Comment (p7nov) [inline docs/topics/jvm/jvm-records.md:27]:
```suggestion
To access the record component, just use its name like you do for [Kotlin properties](properties.md):
```

**Review by p7nov — APPROVED** [review]
LGTM
Just a couple of small suggestions.

File: docs/topics/jvm/jvm-records.md (hunk 11 above)
Comment (koshachy) [reply docs/topics/jvm/jvm-records.md:16]: That one is tricky – I usually try not to use more than one logical markup element. I think there is room for discussion :)
However, I'll approve your suggestion.

**Pull request merged**